grafana/terraform-provider-grafana
# Add support for metrics endpoint scrape jobs

**What:** This PR adds resources to manage ["metrics endpoint scrape jobs" ](https://grafana.com/docs/grafana-cloud/monitor-infrastructure/integrations/integration-reference/integration-metrics-endpoint/).
Metrics Endpoint enables customers to configure Grafana Cloud managed agents to scrape their publicly addressable metrics endpoint.

**Why:** Setting up these scrape jobs has only been available in the Grafana Cloud UI until now, and this PR provides a solution to manage them via Terraform. ([Design doc](https://docs.google.com/document/d/1YVThjOtBjV0L7IvSlnCjyOVtmCOttLOhCYhiOsDwaps/edit))

**Special notes:** This branch is composed of several separate PRs that have been reviewed individually. It is based on similar work for [AWS scrape jobs](https://github.com/grafana/terraform-provider-grafana/tree/tf-cloud-provider-aws-app). Thank you, Tristan and Pablo.

#### Testing locally

Prerequisite: Follow the [instructions](https://github.com/grafana/terraform-provider-grafana?tab=readme-ov-file#local-development-with-grafana) for local development

1. Create a directory with two files:
* main.tf with the following content
  ```hcl
  provider "grafana" {
    connections_api_url = "<your connections url>"
    connections_api_access_token = "<your access policy token>"
  }
  
  resource "grafana_connections_metrics_endpoint_scrape_job" "test" {
     stack_id = "<your stack id>"
     name = "my_scrape_job"
     enabled = true
     authentication_method = "basic" # or use bearer and authentication_bearer_token
     authentication_basic_username = "<your username>"
     authentication_basic_password = "<your password>"
     url = "<url to scrape>"
     scrape_interval_seconds = 120
   }
  ```
* terraform.tf with the following content
   ```
   terraform {
     required_providers {
       grafana = {
         source = "grafana/grafana"
       }
     }
   }
   ```

2. Run `terraform apply`
3. Navigate to the Metrics Endpoint connection: you should see a successful scrape job created:
![Screenshot 2024-10-29 at 11 01 51](https://github.com/user-attachments/assets/13094e6b-a218-4c57-a19f-1ab208600d21)
